File: src/graph/relationship.ts
import ASTNode from "../parsing/ast_node";
import Expression from "../parsing/expressions/expression";
import Hops from "./hops";
import Node from "./node";
import RelationshipData, { RelationshipRecord } from "./relationship_data";
import RelationshipMatchCollector, {
    RelationshipMatchRecord,
} from "./relationship_match_collector";

class Relationship extends ASTNode {
    protected _identifier: string | null = null;
    protected _types: string[] = [];
    protected _properties: Map<string, Expression> = new Map();
    protected _hops: Hops = new Hops();

    protected _value: RelationshipMatchRecord | RelationshipMatchRecord[] | null = null;
    protected _matches: RelationshipMatchCollector = new RelationshipMatchCollector();

    protected _source: Node | null = null;
    protected _target: Node | null = null;
    protected _direction: "left" | "right" = "right";

    private _data: RelationshipData | null = null;

    constructor(identifier: string | null = null, type: string | null = null) {
        super();
        this._identifier = identifier;
        if (type !== null) {
            this._types = [type];
        }
    }
    public set identifier(identifier: string) {
        this._identifier = identifier;
    }
    public get identifier(): string | null {
        return this._identifier;
    }
    public set type(type: string) {
        this._types = [type];
    }
    public get type(): string | null {
        return this._types.length > 0 ? this._types[0] : null;
    }
    public set types(types: string[]) {
        this._types = types;
    }
    public get types(): string[] {
        return this._types;
    }
    public get properties(): Map<string, Expression> {
        return this._properties;
    }
    public set properties(properties: Map<string, Expression>) {
        this._properties = properties;
    }
    private _matchesProperties(hop: number = 0): boolean {
        const data: RelationshipData = this._data!;
        for (const [key, expression] of this._properties) {
            const record: RelationshipRecord = data.current(hop)!;
            if (record === null) {
                throw new Error("No current relationship data available");
            }
            if (!(key in record)) {
                throw new Error("Relationship does not have property");
            }
            return record[key] === expression.value();
        }
        return true;
    }
    public setProperty(key: string, value: Expression): void {
        this._properties.set(key, value);
    }
    public getProperty(key: string): Expression | null {
        return this._properties.get(key) || null;
    }
    public set hops(hops: Hops) {
        this._hops = hops;
    }
    public get hops(): Hops | null {
        return this._hops;
    }
    public setValue(relationship: Relationship, traversalId: string = ""): void {
        const match: RelationshipMatchRecord = this._matches.push(relationship, traversalId);
        this._value = this._matches.value();
    }
    public set source(node: Node | null) {
        this._source = node;
    }
    public get source(): Node | null {
        return this._source;
    }
    public set target(node: Node | null) {
        this._target = node;
    }
    public get target(): Node | null {
        return this._target;
    }
    public set direction(direction: "left" | "right") {
        this._direction = direction;
    }
    public get direction(): "left" | "right" {
        return this._direction;
    }
    public value(): RelationshipMatchRecord | RelationshipMatchRecord[] | null {
        return this._value;
    }
    public get matches(): RelationshipMatchRecord[] {
        return this._matches.matches;
    }
    public setData(data: RelationshipData | null): void {
        this._data = data;
    }
    public getData(): RelationshipData | null {
        return this._data;
    }
    public setEndNode(node: Node): void {
        this._matches.endNode = node;
    }
    public _left_id_or_right_id(): string {
        return this._direction === "left" ? "left_id" : "right_id";
    }
    public async find(left_id: string, hop: number = 0): Promise<void> {
        // Save original source node
        const original = this._source;
        if (hop > 0) {
            // For hops greater than 0, the source becomes the target of the previous hop
            this._source = this._target;
        }
        if (hop === 0) {
            this._data?.reset();

            // Handle zero-hop case: when min is 0 on a variable-length relationship,
            // match source node as target (no traversal)
            if (this.hops?.multi() && this.hops.min === 0 && this._target) {
                // For zero-hop, target finds the same node as source (left_id)
                // No relationship match is pushed since no edge is traversed
                await this._target.find(left_id, hop);
            }
        }
        while (this._data!.find(left_id, hop, this._direction)) {
            const data: RelationshipRecord = this._data?.current(hop) as RelationshipRecord;
            const id = data[this._left_id_or_right_id()];
            if (hop + 1 >= this.hops!.min) {
                this.setValue(this, left_id);
                if (!this._matchesProperties(hop)) {
                    continue;
                }
                await this._target?.find(id, hop);
                if (hop + 1 < this.hops!.max) {
                    if (this._matches.isCircular(id)) {
                        this._matches.pop();
                        continue;
                    }
                    await this.find(id, hop + 1);
                }
                this._matches.pop();
            } else {
                // Below minimum hops: traverse the edge without yielding a match
                await this.find(id, hop + 1);
            }
        }
        // Restore original source node
        this._source = original;
    }
}

export default Relationship;

File: flowquery-py/src/graph/relationship.py
"""Graph relationship representation for FlowQuery."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union

from ..parsing.ast_node import ASTNode
from .hops import Hops
from .relationship_data import RelationshipData
from .relationship_match_collector import RelationshipMatchCollector, RelationshipMatchRecord

if TYPE_CHECKING:
    from .node import Node


class Relationship(ASTNode):
    """Represents a relationship in a graph pattern."""

    def __init__(self) -> None:
        super().__init__()
        self._identifier: Optional[str] = None
        self._type: Optional[str] = None
        self._hops: Hops = Hops()
        self._source: Optional['Node'] = None
        self._target: Optional['Node'] = None
        self._direction: str = "right"
        self._data: Optional['RelationshipData'] = None
        self._value: Optional[Union[RelationshipMatchRecord, List[RelationshipMatchRecord]]] = None
        self._matches: RelationshipMatchCollector = RelationshipMatchCollector()
        self._properties: Dict[str, Any] = {}

    @property
    def identifier(self) -> Optional[str]:
        return self._identifier

    @identifier.setter
    def identifier(self, value: str) -> None:
        self._identifier = value

    @property
    def type(self) -> Optional[str]:
        return self._type

    @type.setter
    def type(self, value: str) -> None:
        self._type = value

    @property
    def hops(self) -> Hops:
        return self._hops

    @hops.setter
    def hops(self, value: Hops) -> None:
        self._hops = value

    @property
    def properties(self) -> Dict[str, Any]:
        return self._properties

    @properties.setter
    def properties(self, value: Dict[str, Any]) -> None:
        self._properties = value

    def _matches_properties(self, hop: int = 0) -> bool:
        """Check if current record matches all constraint properties."""
        if not self._properties:
            return True
        if self._data is None:
            return True
        for key, expression in self._properties.items():
            record = self._data.current(hop)
            if record is None:
                raise ValueError("No current relationship data available")
            if key not in record:
                raise ValueError("Relationship does not have property")
            return bool(record[key] == expression.value())
        return True

    @property
    def source(self) -> Optional['Node']:
        return self._source

    @source.setter
    def source(self, value: 'Node') -> None:
        self._source = value

    @property
    def target(self) -> Optional['Node']:
        return self._target

    @target.setter
    def target(self, value: 'Node') -> None:
        self._target = value

    @property
    def direction(self) -> str:
        return self._direction

    @direction.setter
    def direction(self, value: str) -> None:
        self._direction = value

    # Keep start/end aliases for backward compatibility
    @property
    def start(self) -> Optional['Node']:
        return self._source

    @start.setter
    def start(self, value: 'Node') -> None:
        self._source = value

    @property
    def end(self) -> Optional['Node']:
        return self._target

    @end.setter
    def end(self, value: 'Node') -> None:
        self._target = value

    def set_data(self, data: Optional['RelationshipData']) -> None:
        self._data = data

    def get_data(self) -> Optional['RelationshipData']:
        return self._data

    def set_value(self, relationship: 'Relationship') -> None:
        """Set value by pushing match to collector."""
        self._matches.push(relationship)
        self._value = self._matches.value()

    def value(self) -> Optional[Union[RelationshipMatchRecord, List[RelationshipMatchRecord]]]:
        return self._value

    @property
    def matches(self) -> List[RelationshipMatchRecord]:
        return self._matches.matches

    def set_end_node(self, node: 'Node') -> None:
        """Set the end node for the current match."""
        self._matches.end_node = node

    async def find(self, left_id: str, hop: int = 0) -> None:
        """Find relationships starting from the given node ID."""
        # Save original source node
        original = self._source
        is_left = self._direction == "left"
        if hop > 0:
            # For hops greater than 0, the source becomes the target of the previous hop
            self._source = self._target
        if hop == 0:
            if self._data:
                self._data.reset()

            # Handle zero-hop case: when min is 0 on a variable-length relationship,
            # match source node as target (no traversal)
            if self._hops and self._hops.multi() and self._hops.min == 0 and self._target:
                # For zero-hop, target finds the same node as source (left_id)
                # No relationship match is pushed since no edge is traversed
                await self._target.find(left_id, hop)

        def find_match(id_: str, h: int) -> bool:
            if self._data is None:
                return False
            if is_left:
                return self._data.find_reverse(id_, h)
            return self._data.find(id_, h)
        follow_id = 'left_id' if is_left else 'right_id'
        while self._data and find_match(left_id, hop):
            data = self._data.current(hop)
            if data and self._hops and hop + 1 >= self._hops.min:
                self.set_value(self)
                if not self._matches_properties(hop):
                    continue
                if self._target and follow_id in data:
                    await self._target.find(data[follow_id], hop)
                if self._matches.is_circular():
                    raise ValueError("Circular relationship detected")
                if self._hops and hop + 1 < self._hops.max:
                    await self.find(data[follow_id], hop + 1)
                self._matches.pop()
            elif data and self._hops:
                # Below minimum hops: traverse the edge without yielding a match
                if follow_id in data:
                    await self.find(data[follow_id], hop + 1)

        # Restore original source node
        self._source = original
